File: dotnet/xmsclient/Client/Handler/BasicReturnMethodHandler.cs
using System;
using jpmorgan.mina.common;
using log4net;
using OpenAMQ.Framing;
using OpenAMQ.XMS.Client.Message;
using OpenAMQ.XMS.Client.Protocol;
using OpenAMQ.XMS.Client.State;

namespace OpenAMQ.XMS.Client.Handler
{
    public class BasicReturnMethodHandler : IStateAwareMethodListener
    {
        private static readonly ILog _logger = LogManager.GetLogger(typeof(BasicReturnMethodHandler));

        private static readonly BasicReturnMethodHandler _instance = new BasicReturnMethodHandler();

        public static BasicReturnMethodHandler GetInstance()
        {
            return _instance;
        }

        public void MethodReceived(AMQStateManager stateManager, AMQMethodEvent evt) 
        {
            _logger.Debug("New JmsBounce method received");
            UnprocessedMessage msg = new UnprocessedMessage();
            msg.DeliverBody = null;
            msg.BounceBody = (BasicReturnBody) evt.Method;
            msg.ChannelId = evt.ChannelId;

            evt.ProtocolSession.UnprocessedMessageReceived(msg);
        }
    }
}

File: java/src/org/openamq/client/handler/BasicReturnMethodHandler.java
package org.openamq.client.handler;

import org.apache.log4j.Logger;
import org.openamq.AMQException;
import org.openamq.client.message.UnprocessedMessage;
import org.openamq.client.protocol.AMQMethodEvent;
import org.openamq.client.state.AMQStateManager;
import org.openamq.client.state.StateAwareMethodListener;
import org.openamq.framing.BasicReturnBody;

public class BasicReturnMethodHandler implements StateAwareMethodListener
{
    private static final Logger _logger = Logger.getLogger(BasicReturnMethodHandler.class);

    private static final BasicReturnMethodHandler _instance = new BasicReturnMethodHandler();

    public static BasicReturnMethodHandler getInstance()
    {
        return _instance;
    }

    public void methodReceived(AMQStateManager stateManager, AMQMethodEvent evt) throws AMQException
    {
        _logger.debug("New JmsBounce method received");
        final UnprocessedMessage msg = new UnprocessedMessage();
        msg.deliverBody = null;
        msg.bounceBody = (BasicReturnBody) evt.getMethod();
        msg.channelId = evt.getChannelId();

        evt.getProtocolSession().unprocessedMessageReceived(msg);
    }
}
